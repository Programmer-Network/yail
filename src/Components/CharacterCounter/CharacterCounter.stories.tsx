import type { Meta, StoryObj } from "@storybook/react-vite";

import CharacterCounter from ".";

const meta = {
  title: "Core/CharacterCounter",
  component: CharacterCounter,
  tags: ["autodocs"],
  argTypes: {}
} satisfies Meta<typeof CharacterCounter>;

export default meta;
type Story = StoryObj<typeof meta>;

export const ExceededMaxLength: Story = {
  args: {
    text: "lorem ipsum dolor sit amet consectetur adipiscing elit",
    max: 10
  }
};

export const WithinAllowedRange: Story = {
  args: {
    text: "lorem",
    max: 10
  }
};
