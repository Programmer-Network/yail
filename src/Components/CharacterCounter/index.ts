export { default } from "./CharacterCounter";
export * from "./types";
