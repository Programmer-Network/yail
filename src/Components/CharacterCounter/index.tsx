import classNames from "classnames";
import { FC } from "react";

import { ICharacterCounterProps } from "./types";

const CharacterCounter: FC<ICharacterCounterProps> = (
  { text, max } = {
    text: "",
    max: 0
  }
) => {
  return (
    <div
      className={classNames("yl:flex yl:gap-1", {
        "yl:text-text": text.length <= max,
        "yl:text-error": text.length > max
      })}
    >
      <span data-testid='current-length'>{text.length}</span>
      <span data-testid='separator'>/</span>
      <span data-testid='max-length'>{max}</span>
    </div>
  );
};

export default CharacterCounter;
