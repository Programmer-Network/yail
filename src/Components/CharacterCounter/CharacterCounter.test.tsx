import { render, screen } from "@testing-library/react";

import CharacterCounter from "./";

describe("CharacterCounter component", () => {
  test("renders correctly - snapshot test", () => {
    const { asFragment } = render(<CharacterCounter text='Test' max={10} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("displays the correct character count and maximum", () => {
    render(<CharacterCounter text='Hello' max={10} />);
    expect(screen.getByText("Hello".length)).toBeInTheDocument();
    expect(screen.getByText("/")).toBeInTheDocument();
    expect(screen.getByText(10)).toBeInTheDocument();
  });

  test("text color changes when exceeding maximum characters", () => {
    const text = "Hello World";
    const max = 10;
    render(<CharacterCounter text={text} max={max} />);

    expect(screen.getByTestId("current-length")).toHaveTextContent(
      text.length.toString()
    );
    expect(screen.getByTestId("separator")).toHaveTextContent("/");
    expect(screen.getByTestId("max-length")).toHaveTextContent(max.toString());
    expect(screen.getByTestId("current-length").parentNode).toHaveClass(
      "text-error"
    );
  });

  test("text color does not change when within maximum characters", () => {
    render(<CharacterCounter text='Hello' max={10} />);
    const counterElement = screen.getByText("Hello".length);
    expect(counterElement).not.toHaveClass("!lg-text-error");
  });
});
