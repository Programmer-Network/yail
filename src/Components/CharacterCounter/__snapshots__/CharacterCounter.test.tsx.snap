// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterCounter component > renders correctly - snapshot test 1`] = `
<DocumentFragment>
  <div
    class="flex gap-1 text-text"
  >
    <span
      data-testid="current-length"
    >
      4
    </span>
    <span
      data-testid="separator"
    >
      /
    </span>
    <span
      data-testid="max-length"
    >
      10
    </span>
  </div>
</DocumentFragment>
`;
